### PR TITLE
Ark jarslink

### PR DESCRIPTION
**Pull request by QilongZhang**
### Motivation:

Transform Ark to support jarslink2.0

### Modification:

Transform Ark to support jarslink2.0

### Result:

Transform Ark to support jarslink2.0

### CODE REVIEW

**Comment by QilongZhang** [comment]
https://github.com/alipay/sofa-ark/issues/59

File: sofa-ark-parent/core/common/src/main/java/com/alipay/sofa/ark/common/util/ThreadPoolUtils.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+     * Build Queue
+     *
+     * @param size size of queue
+     * @param isPriority whether use priority queue or not
+     * @return queue
+     */
+    public static BlockingQueue<Runnable> buildQueue(int size, boolean isPriority) {
+        BlockingQueue<Runnable> queue;
+        if (size == 0) {
+            queue = new SynchronousQueue<Runnable>();
```
Comment (straybirdzls) [inline sofa-ark-parent/core/common/src/main/java/com/alipay/sofa/ark/common/util/ThreadPoolUtils.java]: Can be simplified to `queue = new SynchronousQueue<>()` ; And can do code scan to check whether there are any other code style problems

File: sofa-ark-parent/core/common/src/main/java/com/alipay/sofa/ark/common/util/OrderComparator.java [hunk 2]
```diff
@@ -14,30 +14,24 @@
...
-            } else {
-                return 1;
-            }
-        }
-
-        return Integer.compare(o1.getServiceProviderPriority(), o2.getServiceProviderPriority());
+    public int compare(PriorityOrdered o1, PriorityOrdered o2) {
+        int order1 = o1.getPriority();
+        int order2 = o2.getPriority();
+        return Integer.compare(order1, order2);
```
Comment (straybirdzls) [inline sofa-ark-parent/core/common/src/main/java/com/alipay/sofa/ark/common/util/OrderComparator.java]: I think these 3 lines can be simplified to one line and still very clear

File: sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/Biz.java [hunk 3]
```diff
@@ -17,6 +17,7 @@
 package com.alipay.sofa.ark.spi.model;
 
 import com.alipay.sofa.ark.exception.ArkException;
```
Comment (straybirdzls) [inline sofa-ark-parent/core/spi/src/main/java/com/alipay/sofa/ark/spi/model/Biz.java]: Ark Exception import is unuseful

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/PluginContextImpl.java [hunk 4]
```diff
@@ -63,23 +64,41 @@ public Plugin getPlugin(String pluginName) {
...
+    public List<ServiceReference> referenceServices(ServiceFilter serviceFilter) {
+        List<ServiceReference> references = registryService
+            .referenceServices(new DefaultServiceFilter()
+                .setProviderType(ServiceProviderType.ARK_PLUGIN));
+        List<ServiceReference> result = new ArrayList<>();
+        for (ServiceReference reference : references) {
+            if (serviceFilter.match(reference)) {
+                result.add(reference);
+            }
+        }
```
Comment (straybirdzls) [inline sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/model/PluginContextImpl.java]: I think this api is not good enough to use. It's a bit hard for user to referenceServices with serviceFilter. When I referenceServices, I first focus on interfaceType, then uniqueId and final the filter. Although filter contains the previous 2 elements, but it's not a friendly api from my opinion. And add `@suppressWarning("unchecked")` here

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/biz/BizManagerServiceImpl.java [hunk 5]
```diff
@@ -33,34 +38,137 @@
...
+
+    @Override
+    public Biz unRegisterBizStrictly(String bizName, String bizVersion) {
+        AssertUtils.isFalse(StringUtils.isEmpty(bizName), "Biz name must not be empty.");
+        AssertUtils.isFalse(StringUtils.isEmpty(bizVersion), "Biz version must not be empty.");
+        ConcurrentHashMap<String, Biz> bizCache = bizRegistration.get(bizName);
+        if (bizCache != null) {
+            return bizCache.remove(bizVersion);
         }
+        return null;
```
Comment (straybirdzls) [inline sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/biz/BizManagerServiceImpl.java]: Add comments on the api about the meaning of strictly

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/biz/DefaultBizDeployer.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import com.alipay.sofa.ark.common.log.ArkLogger;
+import com.alipay.sofa.ark.common.log.ArkLoggerFactory;
+import com.alipay.sofa.ark.container.model.BizModel;
+import com.alipay.sofa.ark.exception.ArkException;
+import com.alipay.sofa.ark.spi.model.Biz;
+import com.alipay.sofa.ark.spi.model.BizState;
+import com.alipay.sofa.ark.spi.service.ArkInject;
+import com.alipay.sofa.ark.spi.service.biz.BizDeployer;
+import com.alipay.sofa.ark.spi.service.biz.BizManagerService;
+
```
Comment (straybirdzls) [inline sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/service/biz/DefaultBizDeployer.java]: two unuseful imports here

File: sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/pipeline/SystemPropertiesSettingStage.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+ * @author qilong.zql
+ * @since 0.4.0
+ */
+@Singleton
+public class SystemPropertiesSettingStage implements PipelineStage {
+
+    @Override
+    public void process(PipelineContext pipelineContext) throws ArkException {
+        EnvironmentUtils.setSystemProperty(Constants.SPRING_BOOT_ENDPOINTS_JMX_ENABLED,
+            String.valueOf(false));
```
[the remark this answers is not in the document]
Comment (QilongZhang) [reply sofa-ark-parent/core-impl/container/src/main/java/com/alipay/sofa/ark/container/pipeline/SystemPropertiesSettingStage.java]: done

**Pull request merged**